Buffs Rework

lzardy opened the pull request:
Buffs centralized to Buff class, including scripts and hud elements. 
Objects handle addition/removal logic of buffs, which is tailored to BuffType and BuffAddType.
Buff scripts have been given variables which are required for Buff instantiation.
Many packets required for this PR have been added and implemented.
Many files have had their variables and functions alphabetically ordered (because I thought it looked nice).
Some functions regarding Buffs used by scripts have been added and refined.

Resolve #572

lzardy committed:
Buff Rework (WIP)

Features best explained via experimentation.

lzardy committed:
Better Syntax for Buff Variables

lzardy committed:
Fix Errors

lzardy committed:
Transition Buff Removal to Deactivation

lzardy committed:
Finish

Buffs show their correct time on the hud.
Improved efficiency regarding usage of buff slots.
Make this branch ship-ready.

danil179 commented on Content/LeagueSandbox-Scripts/Buffs/EzrealWBuff/EzrealWBuff.cs:
```diff
@@ -14,30 +14,28 @@ namespace EzrealWBuff
+        {
+            StatsModifier.AttackSpeed.PercentBonus = StatsModifier.AttackSpeed.PercentBonus + (0.15f + 0.05f * ownerSpell.Level);
 
-            _visualBuff = AddBuffHudVisual("EzrealEssenceFluxBuff", 5f, 1, BuffType.COMBAT_ENCHANCER, unit);
```
Visual buffs got deleted with this PR?

lzardy replied:
Not deleted, merged, the packet that was used for Visual Buffs is the just AddBuff packet in this PR.

danil179 commented on GameServerLib/GameObjects/Particle.cs:
```diff
@@ -6,19 +6,26 @@ namespace LeagueSandbox.GameServer.GameObjects
+        public void Remove()
+        {
+            SetToRemove();
+            _game.PacketNotifier.NotifyFXKill(this);
```
I think `SetToRemove` call a function like `OnRemove`. If yes, please move `_game.PacketNotifier.NotifyFXKill(this);` to there (as we don't want to have sync problems).

danil179 commented on GameServerLib/ObjectManager.cs:
```diff
@@ -135,9 +135,20 @@ public void Update(float diff)
+                        var tempbuff = tempBuffs[i];
+                        if (tempbuff.Elapsed())
+                        {
+                            if (ai.HasBuff(tempbuff))
```
What are the scenarios that `ai` doesn't have the buff? Also, please merge the `ifs` into one `if`. ( `if (tempbuff.Elapsed() && ai.HasBuff(tempbuff))` if it works)

The code in the loop seems to be a little bit of spaghetti (some `continue` and ifs that most of the time never occur and correlated with others).

lzardy replied:
The continue is necessary as it prevents elapsed buffs from Updating, which would cause an out of range indexing error.

danil179 reviewed (commented):
Some comments. I still have 5 files left to review.

lzardy committed:
Resolve Particle Removal

Adds OnRemoved  override to Particle class and moves packet notification to it.

lzardy committed:
Resolve Natural Removals

Removed an unnecessary if statement and an unnecessary BuffList removal.

danil179 commented on GameServerLib/ObjectManager.cs:
```diff
@@ -135,9 +135,15 @@ public void Update(float diff)
+                            i--;
+                            continue;
+                        }
+                        tempBuffs[i].Update(diff);
```
Instead of the current loop, try to write these 2 more readable lines:
`tempBuffs.Where(b => b.Elapsed()).ForEach(b => ai.RemoveBuff(b));`
`tempBuffs.Where(b => !b.Elapsed()).ForEach(b => b.Update());`

lzardy replied:
Isn't the current method more efficient? It only uses one for loop compared to your suggestion of two. I'm not familiar with how efficient the List.ForEach is, so I'm just verifying.

danil179 replied:
Please fix this to start from the end (as we discussed) to avoid the changing of `i` inside the loop. My suggested solution I commented here is completely wrong.

danil179 commented on GameServerLib/GameObjects/Spells/Buff.cs:
```diff
@@ -10,91 +10,147 @@ namespace LeagueSandbox.GameServer.GameObjects.Spells
 {
     public class Buff : IBuff
```
I think it should implement `IStackable`.

EDIT:

I see that `IBuff` is inheriting `IStackable` already. We probably want an abstract class `Stackable` that will implement this interface and simply inherit it instead of implementing it everywhere needed.

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs:
```diff
@@ -98,140 +96,147 @@ public class ObjAiBase : AttackableUnit, IObjAiBase
-        }
+                if (!Buffs.ContainsKey(b.Name))
+                {
+                    if (HasBuff(b.Name))
```
How is this if ever results with true? Please explain.

lzardy replied:
Buffs is the dictionary of parent buffs. HasBuff checks the BuffList, which is a list of all buffs, not just parent buffs. 
The if statement for HasBuff returns true when the unit has a buff of the same name, but no parent buff for buffs of that name.
This if statement is only true for STACKS_AND_OVERLAPS BuffAddType buffs.

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs:
```diff
@@ -632,6 +552,147 @@ public bool RecalculateAttackPosition()
+        {
+            lock (_buffsLock)
+            {
+                for (int i = 0; i < BuffList.Count; i++)
```
Use LINQ statement like `BuffList.Where(x => x.IsBuffSame(buffname)).ForEach(x => x.DeactivateBuff())` instead of for.

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs:
```diff
@@ -389,175 +397,87 @@ private byte GetBuffSlot(IBuff buffToLookFor = null)
+
+        public bool HasBuff(IBuff buff)
+        {
+            foreach (var b in BuffList)
```
Replace this `foreach` with LINQ statement.

danil179 commented on GameServerLib/GameObjects/AttackableUnits/AI/ObjAIBase.cs:
```diff
@@ -389,175 +397,87 @@ private byte GetBuffSlot(IBuff buffToLookFor = null)
+        public bool HasBuff(string buffName)
         {
-            foreach (var cc in _crowdControlList)
+            foreach (var b in BuffList)
```
Replace this `foreach` with LINQ statement.

danil179 reviewed (changes requested):
There is still some code flow that I think introduced unnecessary complexity.

lzardy committed:
Implement Abstract Stackable Class

lzardy committed:
Use LINQ for Buff foreach statements in ObjAiBase

danil179 reviewed (commented):
All the changes seem perfect. I'll approve & merge after the small fix needed in the loop of `ObjectManager.cs`.

danil179 commented:
Also, you need to update your merge as the build currently fail. I suggest to only use rebase if possible.

lzardy committed:
Finish Code Revisions

Implemented reverse for loop in ObjectManager for natural Buff removals.
Cleaned ObjAiBase.HasBuff function.
Added exception call to Buff class for when a Stackable Buff is instantiated with a MaxStacks less than 2.

lzardy commented:
Rebase successful, final revisions commited, ready to merge.
Close #572.
